starboard/rdk/launcher.py

```diff
@@ -0,0 +1,426 @@
+
+  _SSH_LOGIN_SIGNAL = 'cobalt-launcher-login-success'
+  _SSH_SLEEP_SIGNAL = 'cobalt-launcher-done-sleeping'
+
+  # pexpect times out each second to allow Kill to quickly stop a test run
+  _PEXPECT_TIMEOUT = 1
+
+  # SSH shell command retries
+  _PEXPECT_SPAWN_RETRIES = 20
+
+  # pexpect.sendline retries
+  _PEXPECT_SENDLINE_RETRIES = 3
+
+  # Old process kill retries
+  _KILL_RETRIES = 3
+
+  _PEXPECT_SHUTDOWN_SLEEP_TIME = 3
+  # Time to wait after processes were killed
+  _PROCESS_KILL_SLEEP_TIME = 10
+
+  # Retrys for getting a clean prompt
+  _PROMPT_WAIT_MAX_RETRIES = 5
+  # Wait up to 10 seconds for the password prompt from the RDK
+  _PEXPECT_PASSWORD_TIMEOUT_MAX_RETRIES = 10
+  # Wait up to 900 seconds for new output from the RDK
+  _PEXPECT_READLINE_TIMEOUT_MAX_RETRIES = 900
+  # Delay between subsequent SSH commands
+  _INTER_COMMAND_DELAY_SECONDS = 1.5
+
+  # This is used to strip ansi color codes from pexpect output.
+  _PEXPECT_SANITIZE_LINE_RE = re.compile(r'\x1b[^m]*m')
+
+  # Exceptions to retry
+  _RETRY_EXCEPTIONS = (pexpect.TIMEOUT, pexpect.ExceptionPexpect,
+                       pexpect.exceptions.EOF, OSError)
+
+  def __init__(self, platform, target_name, config, device_id, **kwargs):
+    # pylint: disable=super-with-arguments
+    super().__init__(platform, target_name, config, device_id, **kwargs)
+    env = os.environ.copy()
+    env.update(self.env_variables)
+    self.full_env = env
+    self.platform = platform
+
+    if not self.device_id:
+      self.device_id = self.full_env.get('RDK_ADDR')
+      if not self.device_id:
+        raise ValueError(
+            'Unable to determine target, please pass it in, or set RDK_ADDR '
+            'environment variable.')
+
+    self.startup_timeout_seconds = Launcher._STARTUP_TIMEOUT_SECONDS
+
+    self.pexpect_process = None
+
+    self._InitPexpectCommands()
+
+    self.run_inactive = threading.Event()
+    self.run_inactive.set()
+
+    self.shutdown_initiated = threading.Event()
+
+    self.log_targets = kwargs.get('log_targets', True)
+
+    signal.signal(signal.SIGINT, functools.partial(_sigint_or_sigterm_handler))
+    signal.signal(signal.SIGTERM, functools.partial(_sigint_or_sigterm_handler))
+
+    self.last_run_pexpect_cmd = ''
+
+  def _InitPexpectCommands(self):
+    """Initializes all of the pexpect commands needed for running the test."""
+
+    # Ensure no trailing slashes
+    self.out_directory = self.out_directory.rstrip('/')
+
+    rdk_user_hostname = f'{Launcher._RDK_USERNAME}@{self.device_id}'
+    rdk_test_dir = '/usr/share/content/data/app'
+
+    # scp command setup
+    options = '-rOCq'
+    rsa_options = (
+        '-o \"LogLevel ERROR\" '
+        '-o \"UserKnownHostsFile=/dev/null\" -o \"StrictHostKeyChecking=no\"')
+
+    source = os.path.join(self.out_directory, 'content', 'app', 'cobalt')
+    destination = f'{rdk_user_hostname}:{rdk_test_dir}/'
+    self.scp_command = (f'scp {options} {rsa_options} '
```

Comment:
Why are you using scp instead of rsync to copy over the files? Is rsync not supported on the RDK device?

Reply:
Switched back to rsync.